### ISSUE TITLE
Errors building on Linux

### ISSUE THREAD

**Issue by alexander-g2** [tbr/bjontegaard_cpp#2]
I got errors when trying to build the program on Fedora Linux. I installed conan with `pip install conan` then tried building bdcpp:

```
[fedora@fedora build folder]$ conan install '/home/fedora/Downloads/bjontegaard_cpp-main' -s build_type=Release
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.txt: Installing package
Requirements
    eigen/3.4.0 from 'conancenter' - Cache
Packages
    eigen/3.4.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache

Installing (downloading, building) binaries...
eigen/3.4.0: Already installed!
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Generator cmake_find_package_multi created Eigen3ConfigVersion.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Target-release.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Targets.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Config.cmake
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo
```
and

```
[fedora@fedora build folder]$ cmake '/home/fedora/Downloads/bjontegaard_cpp-main'
-- The CXX compiler identification is GNU 12.0.1
-- The C compiler identification is GNU 12.0.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Could NOT find eigen3 (missing: eigen3_DIR)
-- Configuring done
CMake Error at CMakeLists.txt:66 (add_library):
  Target "bdcpp" links to target "Eigen3::Eigen" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:76 (add_executable):
  Target "example" links to target "Eigen3::Eigen" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Attached resulting build folder:
[build folder.zip](https://github.com/tbr/bjontegaard_cpp/files/9889286/build.folder.zip)

**Comment by tbr** [comment]
Hi, first of all build related issues are a little bit out-of-scope of the project. However, let's see if I can help.

1) Please not that the way you are building (using conan and cmake) is by letting these tools place everything in the root source folder of the project. This is considered a bad practice, as it will mix build-related files and source code. It is difficult to start from clean by mixing everything. A better approach is to make a new folder inside of the '/home/fedora/Downloads/bjontegaard_cpp-main' (in your case), for example a `build` folder. So, create `/home/fedora/Downloads/bjontegaard_cpp-main/build` and cd into that folder.

2) So, your working directory should be the `build` folder. From there issue the following commands:
`conan install .. -s build_type=Release` and `cmake .. -DCMAKE_BUILD_TYPE=Release`. Both of these commands should run without errors.

3) Inside the `build` folder, issue `make` to build the code.

The advantage is that everything related to build is bow contained in the `build` folder. If you want to start clean, just delete the `build` and start over.

Now, I suppose the reason that cmake complains for you is that you are configuring for Debug build, while you only installed the Release version of Eigen. The build type that cmake configures can be specified with the `-DCMAKE_BUILD_TYPE=<...>` option.

If this doesn't solve your issue, I cannot easily help you further, because it is probably a Fedora related thing. Maybe try installing Eigen library via the Fedora system package manager or an alternative way. In principle, the bjontgaard code doesn't care if you use conan or not (conan is just another package manager for C++ builds).

**Comment by tbr** [comment]
Hi, I looked into this via a Fedora (36) docker instance. It seems that Fedora installs conan 2.0 beta, and not conan 1.53. Apparently the new conan is different and requires changes to the files.

Please switch to conan 1.53 (or similar) and everything will magically work. The easiest way to get conan 1.53 is by installing it via Python Pip (`pip install conan`).

I'm not immediately going to fix this (you are welcome to do so) to make the build work with conan 2.0 (and the fact that it breaks compatibility makes me want to rethink using conan at all).

**Comment by alexander-g2** [comment]
Thank you for responding so quickly and for helping me.

Regarding your first suggestion, I was initially creating the build folder outside the source directory, i.e. `./Downloads/build folder/`. My Conan is version 1.53.0.

What helped me was renaming the `Eigen3Config.cmake` file `conan install .. -s build_type=Release` creates to `eigen3Config.cmake` (lower case E) and running `cmake .. -DCMAKE_BUILD_TYPE=Release` instead of `cmake ..` like you suggested.

```
[fedora@fedora build]$ conan install .. -s build_type=Release
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.txt: Installing package
Requirements
    eigen/3.4.0 from 'conancenter' - Cache
Packages
    eigen/3.4.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache

Installing (downloading, building) binaries...
eigen/3.4.0: Already installed!
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Generator cmake_find_package_multi created Eigen3ConfigVersion.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Target-release.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Targets.cmake
conanfile.txt: Generator cmake_find_package_multi created Eigen3Config.cmake
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo
```
```
[fedora@fedora build]$ mv Eigen3Config.cmake eigen3Config.cmake
```
```
[fedora@fedora build]$ cmake .. -DCMAKE_BUILD_TYPE=Release
-- The CXX compiler identification is GNU 12.0.1
-- The C compiler identification is GNU 12.0.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fedora/Downloads/bjontegaard_cpp-main/build
```
```
[fedora@fedora build]$ make
[ 25%] Building CXX object CMakeFiles/bdcpp.dir/bdcpp/bdcpp.cpp.o
[ 50%] Linking CXX static library libbdcpp.a
[ 50%] Built target bdcpp
[ 75%] Building CXX object CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable bin/example
[100%] Built target example
```
But the bin folder only contains the example program. Shouldn't there be bdcpp too?

> I'm not immediately going to fix this (you are welcome to do so) to make the build work with conan 2.0 (and the fact that it breaks compatibility makes me want to rethink using conan at all).

I'm not a computer programmer, I'm afraid I cannot contribute code (I know what am I doing on this website). Simplifying building the program, or providing precompiled binaries, is probably greatly appreciated.

**Comment by tbr** [comment]
Ah, then it might be a bug in the CMakeLists.txt (in `find_package` where it searches for `eigen3` instead of `Eigen3`). I will see if that fixes it.

The code here is delivered as basically two files (a `cpp` and a `h` file) that does everything. Then there is just a short example file on how to use the code. It is not a command line tool or so (also beyond the scope of this project). For you to use the functions, you should take the code and place it into a project of your own.

If you are looking for easier Bjontegaard calculations, there is also the https://github.com/tbr/bjontegaard_etro project to do everything in Excel.

**Comment by alexander-g2** [comment]
Ah, ok. That clarifies it. I'll try to understand what the code is doing then and build a program around it. Thanks for the help again.

I'll just leave the Issue for you to close if you want to fix that CMakeLists.txt bug.

**Comment by tbr** [comment]
Fixed in commit #c336cdbf1ea8e0e68eee76bea328efa06e400872
